service/history/queues/rescheduler.go

```diff
@@ -173,49 +176,68 @@ func (r *reschedulerImpl) reschedule() {
 	r.Lock()
 	defer r.Unlock()
 
-	r.metricsHandler.Histogram(TaskReschedulerPendingTasks, metrics.Dimensionless).Record(int64(r.pq.Len()))
-
-	var failToSubmit []rescheduledExecuable
-	for !r.pq.IsEmpty() {
-		if r.timeSource.Now().Before(r.pq.Peek().rescheduleTime) {
-			break
-		}
-
-		rescheduled := r.pq.Remove()
-		executable := rescheduled.executable
-		if executable.State() == ctasks.TaskStateCancelled {
-			continue
-		}
-
-		submitted, err := r.scheduler.TrySubmit(executable)
-		if err != nil {
-			executable.Logger().Error("Failed to reschedule task", tag.Error(err))
+	r.metricsHandler.Histogram(TaskReschedulerPendingTasks, metrics.Dimensionless).Record(int64(r.numExecutables))
+
+	now := r.timeSource.Now()
+	for key, pq := range r.pqMap {
+		for !pq.IsEmpty() {
+			rescheduled := pq.Remove()
+
+			if now.Before(rescheduled.rescheduleTime) {
+				pq.Add(rescheduled)
+				r.timerGate.Update(rescheduled.rescheduleTime)
+				break
+			}
+
+			executable := rescheduled.executable
+			if executable.State() == ctasks.TaskStateCancelled {
+				r.numExecutables--
+				continue
+			}
+
+			submitted := r.scheduler.TrySubmit(executable)
+			if !submitted {
+				pq.Add(rescheduled)
+				r.timerGate.Update(now.Add(taskChanFullBackoff))
+				break
+			}
+
+			r.numExecutables--
 		}
 
-		if !submitted {
-			rescheduled.rescheduleTime.Add(rescheduleFailureBackoff)
-			failToSubmit = append(failToSubmit, rescheduled)
+		if pq.IsEmpty() {
+			delete(r.pqMap, key)
```

Comment:
maybe worth to consider object pooling if profiling shows frequent create/destroy pq become a bottleneck.

Reply:
yeah make sense.

I am also thinking that maybe this check can be moved to the beginning of the loop, so we don't immediately clean up the pq when it's empty but wait for the next round to see if it's still empty. If there's no next round, storing some empty pq seems ok to me?

Reply:
I updated the logic to clean up empty PQ every 3 minutes to avoid keep creating & deleting pq, especially for scheduled queue. 

I am going to land this PR. Let me know if you have any concern re. ^ change. If so, I am happy to change it in a later pr. Thanks.